cli/api/daemon.go

```diff
@@ -1088,6 +1088,63 @@ func (d *daemon) startAgent() error {
 		log.Debug("Registered local Agent RPC service")
 	}
 
+	// ZEN-17361
+	tTicker := time.NewTicker(time.Minute * 5)
+	d.waitGroup.Add(1)
+	go func() {
+		defer func() {
+			tTicker.Stop()
+			d.waitGroup.Done()
+		}()
+		for {
+			select {
+			case <-d.shutdown:
+				log.Info("Stopping RAM threshold reporter")
+				return
+			case <-tTicker.C:
+				getAllServices, err := d.facade.GetAllServices(d.dsContext)
+				if err != nil {
+					log.Errorln(err)
+					continue
+				}
+				for _, getService := range getAllServices {
+					if getService.RAMThreshold == 0 || getService.RAMCommitment.Value <= 0 || getService.CurrentState != string(service.SVCCSRunning) {
+						continue
+					}
+					getServiceInstances, err := d.facade.GetServiceInstances(d.dsContext, time.Now().Add(5*-time.Minute), getService.GetID())
+					if err != nil {
+						log.Errorln(err)
+						continue
+					}
+					ServiceInstances := make([]metrics.ServiceInstance, 0, len(getServiceInstances))
+					for _, instance := range getServiceInstances {
+						ServiceInstances = append(ServiceInstances, metrics.ServiceInstance{instance.ServiceID, instance.InstanceID})
+					}
+					ramMetric, err := d.facade.GetInstanceMemoryStats(time.Now().Add(5*-time.Minute), ServiceInstances...)
+					if err != nil {
+						log.Errorln(err)
+					}
+					for _, insta := range ramMetric {
+						percent := int((float64(insta.Last) * 100) / float64(getService.RAMCommitment.Value))
+						if int(percent) >= int(getService.RAMThreshold) {
+							a, err := strconv.Atoi(insta.InstanceID)
+							if err != nil {
+								log.Errorln(err)
+								continue
+							}
+							if err := d.facade.StopServiceInstance(d.dsContext, insta.ServiceID, a); err != nil {
+								log.Errorln(err)
+							} else {
+								log.Info("Restarted instance by RAM threshold", insta.ServiceID, a)
+							}
+						}
+					}
+				}
+
+			}
+		}
+	}()
```

Comment:
This should only run on the master.  Several other goroutines in this function have checks for whether the running process is master or delegate and this should do the same.  Since this routine looks at all the services, it should only run in one instance, i.e. master.